x-pack/plugins/security/public/session/session_timeout.tsx

```diff
@@ -6,239 +6,367 @@
+  lastExtensionTime: number;
+  expiresInMs: number;
+  canBeExtendedByMs: number;
+}
+
+export class SessionTimeout {
+  private channel?: BroadcastChannel<SessionState>;
+
+  private isVisible = document.visibilityState !== 'hidden';
+  private isExtending = false;
+
+  private sessionState$ = new BehaviorSubject<SessionState>({
+    lastExtensionTime: 0,
+    expiresInMs: 0,
+    canBeExtendedByMs: 0,
+  });
+  private subscription?: Subscription;
+
   private warningToast?: Toast;
 
+  private stopActivityMonitor?: Function;
+  private stopVisibilityMonitor?: Function;
+  private removeHttpInterceptor?: Function;
+
+  private stopRefreshTimer?: Function;
+  private stopWarningTimer?: Function;
+  private stopLogoutTimer?: Function;
+
   constructor(
     private notifications: NotificationsSetup,
     private sessionExpired: ISessionExpired,
     private http: HttpSetup,
     private tenant: string
   ) {}
 
-  start() {
+  public start() {
     if (this.http.anonymousPaths.isAnonymous(window.location.pathname)) {
       return;
     }
 
-    // subscribe to a broadcast channel for session timeout messages
-    // this allows us to synchronize the UX across tabs and avoid repetitive API calls
-    const name = `${this.tenant}/session_timeout`;
-    this.channel = new BroadcastChannel(name, { webWorkerSupport: false });
-    this.channel.onmessage = this.handleSessionInfoAndResetTimers;
+    this.subscription = this.sessionState$
+      .pipe(skip(1), throttleTime(1000), tap(this.toggleEventHandlers))
+      .subscribe(this.resetTimers);
 
-    // Triggers an initial call to the endpoint to get session info;
-    // when that returns, it will set the timeout
-    return this.fetchSessionInfoAndResetTimers();
+    return this.fetchSessionInfo();
   }
 
-  stop() {
-    if (this.channel) {
-      this.channel.close();
-    }
-    this.cleanup();
+  public stop() {
+    const disabled = {
+      lastExtensionTime: 0,
+      expiresInMs: 0,
+      canBeExtendedByMs: 0,
+    };
+    this.toggleEventHandlers(disabled);
+    this.resetTimers(disabled);
+    this.subscription?.unsubscribe();
   }
 
   /**
-   * When the user makes an authenticated, non-system API call, this function is used to check
-   * and see if the session has been extended.
-   * @param url The URL that was called
+   * Event handler that receives session information from other browser tabs.
+   */
+  private handleChannelMessage = (message: SessionState) => {
+    this.sessionState$.next(message);
+  };
+
+  /**
+   * HTTP request interceptor which ensures that API calls extend the session only if tab is
+   * visible.
    */
-  extend(url: string) {
-    // avoid an additional API calls when the user clicks the button on the session idle timeout
-    if (url.endsWith(SESSION_ROUTE)) {
+  private handleHttpRequest = (fetchOptions: HttpFetchOptionsWithPath) => {
+    // Ignore requests to external URLs
+    if (!fetchOptions.path.startsWith('/')) {
       return;
     }
```

Comment:
I don't think we need this, and I think it will actually miss valid internal requests that use relative paths such as `fetch('some/api/call'`.

This request interceptor doesn't handle _all_ HTTP requests (e.g., those made by `window.fetch()`), it only handles HTTP requests which were made using the Core HTTP service's API wrapper (`http.fetch()`). This wrapper can only be used to fetch data from Kibana's backend (see [fetch.ts](https://github.com/elastic/kibana/blob/4584a8b570402aa07832cf3e5b520e5d2cfa7166/src/core/public/http/fetch.ts), [dev docs](https://www.elastic.co/guide/en/kibana/7.13/http-service.html#_client_side_usage)).

Reply:
Unfortunately we do need that condition since the interface is used to fetch the newsfeed from staging/production. 

I'm not sure if any uses relative URLs like `fetch('some/api/call'` since that would be incredibly flaky but I can update the condition to be more restrictive.

Reply:
Ah I wasn't aware this is used for newsfeed, good catch. I guess http.fetch` wasn't _intended_ to be used for third-party URLs but 🤷‍♂️ 

OK, in that case it's probably better to change it to this:
```ts
if (fetchOptions.path.startsWith('http')) {
  return;
}
```

If we went this route, we _could_ wind up setting the `kbn-system-request` header if a plugin used `fetch` with a path that doesn't specify the protocol, like `example.com/foo`, but there's really no harm in that. I'd prefer to take the safest route to limit false negatives so we don't accidentally allow Kibana requests to extend the session in the background when they shouldn't.

Alternatively if you wanted to get really specific you could do something like this:
```ts
import { parse as parseUrl } from 'url';
...
const url = parseUrl(fetchOptions.path, false, true); // slashesDenoteHost: true
if (url.host !== null) {
  return;
}
```

We already use `url` in the security plugin's public code so we won't be increasing the bundle size by using it here. But I'd like to remove the existing usage of that module at some point if possible because it adds quite a bit to the bundle size.